dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+        If None, the gradient is not computed
+    """
+    cdef:
+        double epsilon = 2.2204460492503131e-016
+        double metric_value
+        int nrows = joint_gradient.shape[0]
+        int ncols = joint_gradient.shape[1]
+        int n = joint_gradient.shape[2]
+
+    mi_gradient[:] = 0
+    metric_value = 0
+    for i in range(nrows):
+        for j in range(ncols):
+            if joint[i,j] < epsilon or mmarginal[j] < epsilon:
+                continue
+
+            factor = log(joint[i,j] / mmarginal[j])
+
+            if mi_gradient is not None:
+                for k in range(n):
+                    mi_gradient[k] -= joint_gradient[i, j, k] * factor
+
+            if smarginal[i] > epsilon:
+                metric_value -= joint[i,j] * (factor - log(smarginal[i]))
+
+    return metric_value
+
+
+def sample_domain_2d(int[:] shape, int n, double[:,:] samples,
+                     int[:,:] mask=None):
+    r""" Take n samples from a grid of the given shape where mask is not zero
+
+    The sampling is made without replacement. If the number of grid cells is
+    less than n, then all cells are selected (in random order).
+    Returns the number of samples actually taken
+
+    Parameters
+    ----------
+    shape : array, shape (2,)
+        the shape of the grid to be sampled
+    n : int
+        the number of samples to be acquired
+    samples : array, shape (n, 2)
+        buffer in which to write the acquired samples
+    mask : array, shape (shape[0], shape[1])
+        the mask indicating the valid samples. Only grid cells x with
+        mask[x] != 0 will be selected
+
+    Returns
+    -------
+    m : int
+        the number of samples actually acquired
+
+    Example
+    -------
+    >>> from dipy.align.mattes import sample_domain_2d
+    >>> import dipy.align.vector_fields as vf
+    >>> mask = np.array(vf.create_circle(10, 10, 3), dtype=np.int32)
+    >>> n = 5
+    >>> dim = 2
+    >>> samples = np.empty((n, dim))
+    >>> sample_domain_2d(np.array(mask.shape, dtype=np.int32), n, samples, mask)
+    5
+    >>> [mask[tuple(x)] for x in samples]
+    [1, 1, 1, 1, 1]
+    """
+    cdef:
+        int tmp, m, r, i, j
+        double p, q
+        int[:] index = np.empty(shape=(shape[0]*shape[1], ), dtype=np.int32)
+    with nogil:
+        # make an array of all avalable indices
+        m = 0
+        for i in range(shape[0]):
+            for j in range(shape[1]):
+                if mask is None or mask[i, j] != 0:
+                    index[m] = i * shape[1] + j
+                    m += 1
+        if n > m:
+            n = m
+    selected = np.random.choice(index[:m], n)
+    for i in range(n):
+        samples[i,0] = selected[i] // shape[1]
+        samples[i,1] = selected[i] % shape[1]
+    return n
+
+
+def sample_domain_regular(int k, int[:] shape, double[:,:] affine,
+                          double sigma=0.25, int seed=1234):
```

Comment:
Yes, I think it is better if the sampling is reproducible, otherwise, we may get different results if we run the registration multiple times, which would make it very difficult to evaluate. Do you think we should give the possibility to have more random behavior?